File: Arrays and Strings/stringRotation.js
/**
 * Solution - Append first string to itself and search for second string on it
 *
 * Complexity Analysis
 * O(n) time | O(1) space
 *
 * Checks if str2 is a rotation of str1
 *
 * @param {string} str1
 * @param {string} str2
 * @returns {boolean} True if str1 and str2 are rotated versions of eachother, otherwise false
 */
function isStringRotation(str1 = '', str2 = '') {
    // Gracefully handle type and Falsy values
    if (typeof str1 !== 'string' || typeof str2 !== 'string' || str1 === '' || str2 === '') {
        console.error('Parameter should be a valid non-empty string');
        return false;
    }

    if (str1.length !== str2.length) {
        return false;
    }

    return str1.concat(str1).includes(str2);
}

const testCases = [
    // Normal
    // Data that is typical (expected) and should be accepted by the system.
    { assert: isStringRotation('abcde', 'cdeab'), expected: true },
    { assert: isStringRotation('abcde', 'abced'), expected: false },
    { assert: isStringRotation('waterbottle', 'erbottlewat'), expected: true },

    // Boundary data (extreme data, edge case)
    // Data at the upper or lower limits of expectations that should be accepted by the system.

    // Abnormal data (erroneous data)
    // Data that falls outside of what is acceptable and should be rejected by the system.
    { assert: isStringRotation(), expected: false },
    { assert: isStringRotation(0), expected: false },
    { assert: isStringRotation(''), expected: false },
    { assert: isStringRotation([]), expected: false },
    { assert: isStringRotation({}), expected: false },
    { assert: isStringRotation(NaN), expected: false },
    { assert: isStringRotation(null), expected: false },
    { assert: isStringRotation(false), expected: false },
    { assert: isStringRotation(new Set()), expected: false },
    { assert: isStringRotation(new Map()), expected: false },
    { assert: isStringRotation(undefined), expected: false }
];

// Run tests
testCases.forEach((test, index) => {
    const currentTest = `# Test ${index + 1}`;

    console.group(currentTest);
    console.log('Actual:', test.assert);
    console.log('Expected:', test.expected);
    console.log(test.assert === test.expected ? '🤘 Test PASSED 🤘' : '👎 Test FAILED 👎', '\n');
    console.groupEnd(currentTest);
});

File: Arrays and Strings/stringRotation.py
def isStringRotation(str1='', str2=''):
    '''
    Solution - Append first string to itself and search for second string on it

    Complexity Analysis
    O(n) time | O(1) space

    Checks if str2 is a rotation of str1

    string: str1 
    string: str2
    return: True if str1 and str2 are rotated versions of eachother, otherwise false
    '''
    # Gracefully handle type and Falsy values
    if (not isinstance(str1, str) or not isinstance(str2, str) or str1 == '' or str2 == ''):
        print('Arguments should be a valid non-empty strings')
        return False

    if (len(str1) != len(str2)):
        return False

    return str2 in ''.join((str1, str1))

# Test cases (black box - unit testing)
testCases = [
    # Normal
    # Data that is typical (expected) and should be accepted by the system.
    { 'assert': isStringRotation('abcde', 'cdeab'), 'expected': True },
    { 'assert': isStringRotation('abcde', 'abced'), 'expected': False },
    { 'assert': isStringRotation('waterbottle', 'erbottlewat'), 'expected': True },

    # Boundary data (extreme data, edge case)
    # Data at the upper or lower limits of expectations that should be accepted by the system.

    # Abnormal data (erroneous data)
    # Data that falls outside of what is acceptable and should be rejected by the system.
    { 'assert': isStringRotation(), 'expected': False },
    { 'assert': isStringRotation(0), 'expected': False },
    { 'assert': isStringRotation(''), 'expected': False },
    { 'assert': isStringRotation([]), 'expected': False },
    { 'assert': isStringRotation(()), 'expected': False },
    { 'assert': isStringRotation({}), 'expected': False },
    { 'assert': isStringRotation(None), 'expected': False },
    { 'assert': isStringRotation(False), 'expected': False }
]

# Run tests
for (index, test) in enumerate(testCases):
    print(f'# Test {index + 1}')
    print(f'Actual: {test["assert"]}')
    print(f'expected: {test["expected"]}')
    print('🤘 Test PASSED 🤘' if test["assert"] == test["expected"] else '👎 Test FAILED 👎', '\n')


